### PR TITLE
Fix various damage edge cases

### PR DESCRIPTION
**Pull request by sprunk**
* allow arbitrary armor modifier and damage (incl. 0 and negative)
 * allow edgeEffectiveness == 1 (previously capped at 0.999)
 * fix direct hits sometimes not doing full damage due to fall-off
 * fix some Lua interfaces being able to set edgeEffectiveness above 1

### CODE REVIEW

File: rts/Game/GameHelper.cpp [hunk 1]
```diff
@@ -720,7 +729,7 @@ size_t CGameHelper::GenerateWeaponTargets(const CWeapon* weapon, const CUnit* av
 
 				const float dist2D = math::sqrt(sqDist2D);
 				const float rangeMul = (dist2D * weaponDef->proximityPriority + modRange * 0.4f + 100.0f);
-				const float damageMul = weaponDmg->Get(targetUnit->armorType) * targetUnit->curArmorMultiple;
+				const float damageMul = std::max(0.0001f, weaponDmg->Get(targetUnit->armorType) * targetUnit->curArmorMultiple);
```
[the remark this answers is not in the document]
Comment (sprunk) [reply rts/Game/GameHelper.cpp:732]: There's division by damageMul below. In my tests it wasn't actually a problem but I'd rather be cautious.

File: rts/Sim/Weapons/WeaponDef.cpp [hunk 2]
```diff
@@ -390,7 +386,7 @@ WeaponDef::WeaponDef(const LuaTable& wdTable, const std::string& name_, int id_)
 			const int type = damageArrayHandler.GetTypeFromName(pair.first);
 			if (type == 0)
 				continue;
-			damages.Set(type, std::max(0.0001f, pair.second));
+			damages.Set(type, pair.second);
```
Comment (lhog) [inline rts/Sim/Weapons/WeaponDef.cpp:389]: Consider switching to https://en.cppreference.com/w/cpp/language/structured_binding for loop arguments here.

File: rts/Sim/Weapons/WeaponDef.cpp [hunk 3]
```diff
@@ -386,11 +382,11 @@ WeaponDef::WeaponDef(const LuaTable& wdTable, const std::string& name_, int id_)
 		dmgs.reserve(32);
 		dmgTable.GetPairs(dmgs);
 
-		for (const auto& pair: dmgs) {
-			const int type = damageArrayHandler.GetTypeFromName(pair.first);
-			if (type == 0)
+		for (const auto& [armorTypeName, damage] : dmgs) {
+			const int armorType = damageArrayHandler.GetTypeFromName(armorTypeName);
```
Comment (sprunk) [inline rts/Sim/Weapons/WeaponDef.cpp:386]: Changed to structured binding. The `type` variable was also shadowing `WeaponDef::type` so I changed it to something better as well.

File: rts/Game/GameHelper.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (lhog) [reply rts/Game/GameHelper.cpp:732]: Oh indeed

**Pull request merged**